src/java.base/share/classes/javax/crypto/EncryptedPrivateKeyInfo.java

```diff
@@ -308,6 +320,280 @@ private PKCS8EncodedKeySpec getKeySpecImpl(Key decryptKey,
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+    public PrivateKey getKey(char[] password) throws InvalidKeyException {
+        return getKey(password, null);
+    }
+
+    /**
+     * Returns a {@code PrivateKey} from the encrypted data in this instance.
+     * using the given provider.
+     *
+     * @param password this array is cloned and used for PBE decryption.
+     *                 {@code null} may be used.
+     * @param provider this {@code Provider} is used for all key generations.
+     *                 It is used when generating the PBE encryption key,
+     *                 {@code SecretKeyFactory}, and the {@code PrivateKey},
+     *                 {@code KeyFactory}.  A {@code null} value will use the default
+     *                 provider configuration.
+     * @return a {@code PrivateKey}
+     * @throws InvalidKeyException if an error occurs during parsing of the
+     * encrypted data or creation of the key object.
+     *
+     * @since 25
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+    public PrivateKey getKey(char[] password, Provider provider)
+        throws InvalidKeyException {
+        return getKeyImpl(new PBEKeySpec(password), provider);
+    }
+
+    private PrivateKey getKeyImpl(PBEKeySpec keySpec, Provider provider)
+        throws InvalidKeyException {
+        SecretKeyFactory skf;
+        PKCS8EncodedKeySpec p8KeySpec;
+        try {
+            if (provider == null) {
+                skf = SecretKeyFactory.getInstance(getAlgName());
+                p8KeySpec = getKeySpec(skf.generateSecret(keySpec));
+            } else {
+                skf = SecretKeyFactory.getInstance(getAlgName(), provider);
+                p8KeySpec = getKeySpec(skf.generateSecret(keySpec), provider);
+            }
+
+            return PKCS8Key.parseKey(p8KeySpec.getEncoded(), provider);
+        } catch (GeneralSecurityException e) {
+            throw new InvalidKeyException(e);
+        }
+    }
+
+    /**
+     * Returns a {@code PrivateKey} from the encrypted data in this instance.
+     * using the given provider.
+     *
+     * @param decryptKey this is the decryption key and cannot be {@code null}.
+     * @param provider this {@code Provider} is used for all key generations.
+     *                 It is used when generating the PBE encryption key,
+     *                 {@code SecretKeyFactory}, and the {@code PrivateKey},
+     *                 {@code KeyFactory}.  A {@code null} value will use the
+     *                 default provider configuration.
+     * @return a {@code PrivateKey}
+     * @throws InvalidKeyException if an error occurs during parsing of the
+     * encrypted data or creation of the key object.
+     * @throws NullPointerException if {@code key} is null.
+     *
+     * @since 25
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+    public PrivateKey getKey(Key decryptKey, Provider provider)
```

Comment:
This method goes one step further than the existing `getKeySpec(dk, p)`. It should only throw more types of exception than the existing one. Now you put everything into a `InvalidKeyException`. Is that good?